diff --git a/willuslib/linux.c b/willuslib/linux.c
index 45732ec0..2c28073d 100644
--- a/willuslib/linux.c
+++ b/willuslib/linux.c
@@ -31,7 +31,6 @@
 #include <ctype.h>
 #include <time.h>
 #include <unistd.h>
-#include <sys/termios.h>
 #include <sys/wait.h>
 #include <sys/types.h>
 #include <signal.h>
diff --git a/willuslib/willus.h b/willuslib/willus.h
index 0c4ae1e9..b0f986a4 100644
--- a/willuslib/willus.h
+++ b/willuslib/willus.h
@@ -83,6 +83,10 @@
 
 typedef double  real;
 
+#if defined(__ANDROID__)
+#define ANDROID
+#endif
+
 #if (!defined(CYGWIN32) && defined(__CYGWIN32__))
 #define CYGWIN32
 #endif
@@ -164,7 +168,7 @@ typedef double  real;
 #define WILLUS_BIGENDIAN
 #endif
 
-#if (defined(__GNUC__) && __GNUC__ >= 4 && __GNUC_MINOR__ > 4)
+#if (defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 5)))
 /* For my code */
 #ifndef WILLUS_HAVE_FILE64
 #define WILLUS_HAVE_FILE64
