diff --git a/k2pdfoptlib/k2pdfopt.h b/k2pdfoptlib/k2pdfopt.h
index 5220f482..a87f2784 100644
--- a/k2pdfoptlib/k2pdfopt.h
+++ b/k2pdfoptlib/k2pdfopt.h
@@ -847,7 +847,9 @@ typedef struct
     PAGEREGION *pageregion;
     int n,na;
     } PAGEREGIONS;
+K2PDFOPT_EXPORT
 void pageregions_init(PAGEREGIONS *regions);
+K2PDFOPT_EXPORT
 void pageregions_free(PAGEREGIONS *regions);
 void pageregions_delete_one(PAGEREGIONS *regions,int index);
 void pageregions_insert(PAGEREGIONS *dst,int index,PAGEREGIONS *src);
@@ -974,11 +976,14 @@ int  wrapbmp_remaining(WRAPBMP *wrapbmp,K2PDFOPT_SETTINGS *k2settings);
 void wrapbmp_add(WRAPBMP *wrapbmp,BMPREGION *region,K2PDFOPT_SETTINGS *k2settings,
                  MASTERINFO *masterinfo,int colgap,int justification_flags);
 void wrapbmp_flush(MASTERINFO *masterinfo,K2PDFOPT_SETTINGS *k2settings,int allow_full_justify);
+K2PDFOPT_EXPORT
 void wrectmaps_init(WRECTMAPS *wrectmaps);
+K2PDFOPT_EXPORT
 void wrectmaps_free(WRECTMAPS *wrectmaps);
 void wrectmaps_clear(WRECTMAPS *wrectmaps);
 void wrectmaps_add_wrectmap(WRECTMAPS *wrectmaps,WRECTMAP *wrectmap);
 void wrectmaps_scale_wrapbmp_coords(WRECTMAPS *wrectmaps,double scalew,double scaleh);
+K2PDFOPT_EXPORT
 int  wrectmap_inside(WRECTMAP *wrmap,int xc,int yc);
 void wrectmaps_sort_horizontally(WRECTMAPS *wrectmaps);
 
diff --git a/willuslib/willus.h b/willuslib/willus.h
index b0f986a4..3e2a40a0 100644
--- a/willuslib/willus.h
+++ b/willuslib/willus.h
@@ -446,9 +446,13 @@ void bmp8_to_grey(WILLUSBITMAP *bmap);
 void bmp8_to_bw(WILLUSBITMAP *bmap,int thresh);
 int  bmp8_greylevel_convert(int r,int g,int b);
 #define bmp8_graylevel_convert(r,g,b) bmp8_greylevel_convert(r,g,b)
+K2PDFOPT_EXPORT
 void bmp_init(WILLUSBITMAP *bmap);
+K2PDFOPT_EXPORT
 int  bmp_alloc(WILLUSBITMAP *bmap);
+K2PDFOPT_EXPORT
 int  bmp_bytewidth(WILLUSBITMAP *bmp);
+K2PDFOPT_EXPORT
 unsigned char *bmp_rowptr_from_top(WILLUSBITMAP *bmp,int row);
 void bmp_crop(WILLUSBITMAP *bmp,int x0,int y0_from_top,int width,int height);
 void bmp_crop_ex(WILLUSBITMAP *dst,WILLUSBITMAP *src,int x0,int y0_from_top,int width,int height);
@@ -456,11 +460,13 @@ void bmp_rotate_fast(WILLUSBITMAP *dst,double degrees,int expand);
 int  bmp_rotate_right_angle(WILLUSBITMAP *bmp,int degrees);
 int  bmp_rotate_90(WILLUSBITMAP *bmp);
 int  bmp_rotate_270(WILLUSBITMAP *bmp);
+K2PDFOPT_EXPORT
 int  bmp_copy(WILLUSBITMAP *dest,WILLUSBITMAP *src);
 double bmp_jpeg_bytes_per_pixel(int quality);
 void bmp_flip_horizontal(WILLUSBITMAP *bmp);
 void bmp_flip_vertical(WILLUSBITMAP *bmp);
 int  bmp_bytewidth_win32(WILLUSBITMAP *bmp);
+K2PDFOPT_EXPORT
 void bmp_free(WILLUSBITMAP *bmap);
 int  bmp_read(WILLUSBITMAP *bmap,char *filename,FILE *out);
 void bmp24_reduce_size(WILLUSBITMAP *bmp,int mx,int my);
