diff --git a/willuslib/willus.h b/willuslib/willus.h
index e9f56c49..0c4ae1e9 100644
--- a/willuslib/willus.h
+++ b/willuslib/willus.h
@@ -223,6 +223,7 @@ typedef double  real;
 **
 */
 
+#if 0
 /*
 ** CMAKE handles the defines, not this source
 ** (Mod from Dirk Thierbach, 31-Dec-2013)
@@ -282,6 +283,9 @@ typedef double  real;
 #endif
 
 #endif /* USE_CMAKE */
+#else
+#include "willus_config.h"
+#endif
 /*
 ** Consistency check
 */
