diff --git a/src/linebreak.c b/src/linebreak.c
index 4df7547e..92567efa 100644
--- a/src/linebreak.c
+++ b/src/linebreak.c
@@ -715,6 +715,12 @@ int lb_process_next_char(
 
     return brk;
 }
+enum LineBreakClass lb_get_char_class(
+        struct LineBreakContext *lbpCtx,
+        utf32_t ch)
+{
+    return get_char_lb_class_lang(ch, lbpCtx->lbpLang);
+}
 
 /**
  * Sets the line breaking information for a generic input string.
diff --git a/src/linebreakdef.h b/src/linebreakdef.h
index 2521299e..fe88e417 100644
--- a/src/linebreakdef.h
+++ b/src/linebreakdef.h
@@ -181,5 +181,8 @@ size_t set_linebreaks(
         enum BreakOutputType outputType,
         char *brks,
         get_next_char_t get_next_char);
+enum LineBreakClass lb_get_char_class(
+        struct LineBreakContext *lbpCtx,
+        utf32_t ch);
 
 #endif /* LINEBREAKDEF_H */
