diff --git a/lib/zlib/zconf.h b/lib/zlib/zconf.h
index 5e1d68a0..094a4313 100644
--- a/lib/zlib/zconf.h
+++ b/lib/zlib/zconf.h
@@ -8,6 +8,8 @@
 #ifndef ZCONF_H
 #define ZCONF_H
 
+#define Z_PREFIX
+
 /*
  * If you *really* need a unique prefix for all types and library functions,
  * compile with -DZ_PREFIX. The "standard" zlib should be compiled without it.
@@ -163,6 +165,8 @@
 #  define gz_header_s           z_gz_header_s
 #  define internal_state        z_internal_state
 
+#  define z_errmsg              z_z_errmsg
+
 #endif
 
 #if defined(__MSDOS__) && !defined(MSDOS)
