diff --git a/ccmain/tessedit.cpp b/ccmain/tessedit.cpp
index b2fa0219..41b8c8bf 100644
--- a/ccmain/tessedit.cpp
+++ b/ccmain/tessedit.cpp
@@ -100,6 +100,22 @@ bool Tesseract::init_tesseract_lang_data(
   language_data_path_prefix += lang;
   language_data_path_prefix += ".";
 
+#ifndef NO_CUBE_BUILD
+  int gotcube = 0;
+  if (oem == OEM_DEFAULT) {
+      static const char *exts[] = {"size", "params", "nn", "word-freq", NULL};
+      for (const char **extp = exts; *extp != NULL; ++extp) {
+          STRING cubetest = language_data_path_prefix + "cube." + *extp;
+          FILE *f = fopen(cubetest.string(), "r");
+          if (f != NULL) {
+              gotcube = 1;
+              fclose(f);
+              break;
+          }
+      }
+  }
+#endif
+
   // Initialize TessdataManager.
   STRING tessdata_path = language_data_path_prefix + kTrainedDataSuffix;
   if (!tessdata_manager.Init(tessdata_path.string(),
@@ -157,6 +173,9 @@ bool Tesseract::init_tesseract_lang_data(
 
   // Determine which ocr engine(s) should be loaded and used for recognition.
   if (oem != OEM_DEFAULT) tessedit_ocr_engine_mode.set_value(oem);
+#ifndef NO_CUBE_BUILD
+  else if (gotcube) tessedit_ocr_engine_mode.set_value(OEM_TESSERACT_CUBE_COMBINED);
+#endif
   if (tessdata_manager_debug_level) {
     tprintf("Loading Tesseract/Cube with tessedit_ocr_engine_mode %d\n",
             static_cast<int>(tessedit_ocr_engine_mode));
@@ -207,11 +226,13 @@ bool Tesseract::init_tesseract_lang_data(
   // alternative engines, such as cube and LSTM are optional.
 #ifndef NO_CUBE_BUILD
   if (tessedit_ocr_engine_mode == OEM_CUBE_ONLY) {
-    ASSERT_HOST(init_cube_objects(false, &tessdata_manager));
+    if (!init_cube_objects(false, &tessdata_manager))
+        return false;
     if (tessdata_manager_debug_level)
       tprintf("Loaded Cube w/out combiner\n");
   } else if (tessedit_ocr_engine_mode == OEM_TESSERACT_CUBE_COMBINED) {
-    ASSERT_HOST(init_cube_objects(true, &tessdata_manager));
+    if (!init_cube_objects(true, &tessdata_manager))
+      return false;
     if (tessdata_manager_debug_level)
       tprintf("Loaded Cube with combiner\n");
   }
@@ -320,6 +341,7 @@ int Tesseract::init_tesseract(
       if (!loaded_primary) {
         if (result < 0) {
           tprintf("Failed loading language '%s'\n", lang_str);
+          return -1;
         } else {
           if (tessdata_manager_debug_level)
             tprintf("Loaded language '%s' as main language\n", lang_str);
@@ -331,6 +353,7 @@ int Tesseract::init_tesseract(
         if (result < 0) {
           tprintf("Failed loading language '%s'\n", lang_str);
           delete tess_to_init;
+          return -1;
         } else {
           if (tessdata_manager_debug_level)
             tprintf("Loaded language '%s' as secondary language\n", lang_str);
