diff --git a/gio/meson.build b/gio/meson.build
index 14f0d825..059fec33 100644
--- a/gio/meson.build
+++ b/gio/meson.build
@@ -168,6 +168,7 @@ gnetworking_h_conf.set('NAMESER_COMPAT_INCLUDE', gnetworking_h_nameser_compat_in
 
 gnetworking_h = configure_file(input : 'gnetworking.h.in',
                                output : 'gnetworking.h',
+                               install_tag : 'devel',
                                install_dir : join_paths(get_option('includedir'), 'glib-2.0/gio'),
                                configuration : gnetworking_h_conf)
 
@@ -754,6 +755,7 @@ gioenumtypes_h = custom_target('gioenumtypes_h',
   capture : true,
   input : gio_headers,
   install : true,
+  install_tag : 'devel',
   install_dir : join_paths(get_option('includedir'), 'glib-2.0/gio'),
   command : [python, glib_mkenums,
              '--template', files('gioenumtypes.h.template'),
diff --git a/gobject/meson.build b/gobject/meson.build
index bc44d5db..0c264e62 100644
--- a/gobject/meson.build
+++ b/gobject/meson.build
@@ -106,6 +106,7 @@ glib_enumtypes_h = custom_target('glib_enumtypes_h',
   capture : true,
   input : glib_enumtypes_input_headers,
   install : true,
+  install_tag : 'devel',
   install_dir : join_paths(get_option('includedir'), 'glib-2.0/gobject'),
   command : [python, glib_mkenums,
              '--template', files('glib-enumtypes.h.template'),
