diff --git a/gen.tab/meson.build b/gen.tab/meson.build
index 8c21e98b..72205b6e 100644
--- a/gen.tab/meson.build
+++ b/gen.tab/meson.build
@@ -42,6 +42,7 @@ fribidi_unicode_version_h = custom_target('fribidi-unicode-version.h',
   command: [gen_unicode_version, '@INPUT0@', '@INPUT1@', 'gen-unicode-version'],
   capture: true,
   install_dir: join_paths(get_option('includedir'), 'fribidi'),
+  install_tag: 'devel',
   install: true)
 
 COMPRESSION='2'
diff --git a/lib/meson.build b/lib/meson.build
index fb8f9cf6..160fc17f 100644
--- a/lib/meson.build
+++ b/lib/meson.build
@@ -24,6 +24,7 @@ endif
 fribidi_config_h = configure_file(input: files('fribidi-config.h.in'),
   output: 'fribidi-config.h',
   configuration: fribidi_config,
+  install_tag: 'devel',
   install_dir: join_paths(get_option('includedir'), 'fribidi'))
 
 fribidi_headers = files(
diff --git a/meson.build b/meson.build
index ad2f4e15..e8eddb41 100644
--- a/meson.build
+++ b/meson.build
@@ -1,5 +1,5 @@
 project('fribidi', 'c', version: '1.0.12',
-  meson_version : '>= 0.48')
+  meson_version : '>= 0.60.0')
 
 # New release:
 #   interface_age++
