diff --git a/meson.build b/meson.build
index b80679d3..06082f5b 100644
--- a/meson.build
+++ b/meson.build
@@ -1,5 +1,5 @@
 project('harfbuzz', 'c', 'cpp',
-  meson_version: '>= 0.55.0',
+  meson_version: '>= 0.60.0',
   version: '5.3.1',
   default_options: [
     'cpp_rtti=false',       # Just to support msvc, we are passing -fno-exceptions also anyway
diff --git a/src/meson.build b/src/meson.build
index 2c49007b..006e1d56 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -3,6 +3,7 @@ hb_version_h = configure_file(
                input: 'hb-version.h.in',
                output: 'hb-version.h',
                install: true,
+               install_tag: 'devel',
                install_dir: get_option('includedir') / meson.project_name())
 
 # Base and default-included sources and headers
@@ -665,6 +666,7 @@ configure_file(input: 'harfbuzz-config.cmake.in',
   output: 'harfbuzz-config.cmake',
   configuration: cmake_config,
   install_dir: get_option('libdir') / 'cmake' / 'harfbuzz',
+  install_tag: 'devel',
 )
 
 libharfbuzz_gobject_dep = null_dep
@@ -700,6 +702,7 @@ if have_gobject
     output: 'hb-gobject-enums.h',
     command: [find_program('fix_get_types.py'), '@INPUT@', '@OUTPUT@'],
     install: true,
+    install_tag: 'devel',
     install_dir: get_option('prefix') / get_option('includedir') / meson.project_name(),
   )
 
