diff --git a/src/meson.build b/src/meson.build
index ba3470ff..ff058527 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -531,7 +531,7 @@ libharfbuzz_subset = library('harfbuzz-subset', hb_subset_sources,
   cpp_args: cpp_args + extra_hb_cpp_args,
   soversion: hb_so_version,
   version: version,
-  install: true,
+  install: false,
   darwin_versions: darwin_versions,
   link_language: 'c',
 )
