diff --git a/k2pdfoptlib/k2pdfopt.h b/k2pdfoptlib/k2pdfopt.h
index 7d414dd4..aa5eb797 100644
--- a/k2pdfoptlib/k2pdfopt.h
+++ b/k2pdfoptlib/k2pdfopt.h
@@ -849,7 +849,9 @@ typedef struct
     PAGEREGION *pageregion;
     int n,na;
     } PAGEREGIONS;
+K2PDFOPT_EXPORT
 void pageregions_init(PAGEREGIONS *regions);
+K2PDFOPT_EXPORT
 void pageregions_free(PAGEREGIONS *regions);
 void pageregions_delete_one(PAGEREGIONS *regions,int index);
 void pageregions_insert(PAGEREGIONS *dst,int index,PAGEREGIONS *src);
@@ -976,11 +978,14 @@ int  wrapbmp_remaining(WRAPBMP *wrapbmp,K2PDFOPT_SETTINGS *k2settings);
 void wrapbmp_add(WRAPBMP *wrapbmp,BMPREGION *region,K2PDFOPT_SETTINGS *k2settings,
                  MASTERINFO *masterinfo,int colgap,int justification_flags);
 void wrapbmp_flush(MASTERINFO *masterinfo,K2PDFOPT_SETTINGS *k2settings,int allow_full_justify);
+K2PDFOPT_EXPORT
 void wrectmaps_init(WRECTMAPS *wrectmaps);
+K2PDFOPT_EXPORT
 void wrectmaps_free(WRECTMAPS *wrectmaps);
 void wrectmaps_clear(WRECTMAPS *wrectmaps);
 void wrectmaps_add_wrectmap(WRECTMAPS *wrectmaps,WRECTMAP *wrectmap);
 void wrectmaps_scale_wrapbmp_coords(WRECTMAPS *wrectmaps,double scalew,double scaleh);
+K2PDFOPT_EXPORT
 int  wrectmap_inside(WRECTMAP *wrmap,int xc,int yc);
 void wrectmaps_sort_horizontally(WRECTMAPS *wrectmaps);
 
diff --git a/koptcrop.h b/koptcrop.h
index e2317d37..ad1d8f7e 100644
--- a/koptcrop.h
+++ b/koptcrop.h
@@ -24,6 +24,7 @@
 #include "k2pdfopt.h"
 #include "context.h"
 
+K2PDFOPT_EXPORT
 void k2pdfopt_crop_bmp(KOPTContext *kctx);
 
 #endif
diff --git a/koptimize.h b/koptimize.h
index 6e3daa20..2756e939 100644
--- a/koptimize.h
+++ b/koptimize.h
@@ -25,6 +25,7 @@
 #include "k2pdfopt.h"
 #include "context.h"
 
+K2PDFOPT_EXPORT
 void k2pdfopt_optimize_bmp(KOPTContext *kctx);
 
 #endif
diff --git a/koptocr.c b/koptocr.c
index 7b0fe055..04fa93bc 100644
--- a/koptocr.c
+++ b/koptocr.c
@@ -31,6 +31,7 @@
 #include "setting.h"
 #include "koptocr.h"
 
+K2PDFOPT_EXPORT
 PIX* bitmap2pix(WILLUSBITMAP *src, int x, int y, int w, int h);
 l_int32 k2pdfopt_pixGetWordBoxesInTextlines(PIX *pixs, l_int32 maxsize,
 		l_int32 reduction, l_int32 minwidth, l_int32 minheight,
diff --git a/koptocr.h b/koptocr.h
index 30bd933b..8355c461 100644
--- a/koptocr.h
+++ b/koptocr.h
@@ -28,19 +28,23 @@
 
 void k2pdfopt_tocr_init(char *datadir, char *lang);
 
+K2PDFOPT_EXPORT
 void k2pdfopt_tocr_end();
 
 const char* k2pdfopt_tocr_get_language();
 
+K2PDFOPT_EXPORT
 void k2pdfopt_tocr_single_word(WILLUSBITMAP *src,
 		int x, int y, int w, int h,
 		char *word, int max_length,
 		char *datadir, char *lang, int ocr_type,
 		int allow_spaces, int std_proc);
 
+K2PDFOPT_EXPORT
 void k2pdfopt_get_reflowed_word_boxes(KOPTContext *kctx, WILLUSBITMAP *src,
 		int x, int y, int w, int h);
 
+K2PDFOPT_EXPORT
 void k2pdfopt_get_native_word_boxes(KOPTContext *kctx, WILLUSBITMAP *src,
         int x, int y, int w, int h);
 
diff --git a/koptreflow.h b/koptreflow.h
index 57cd7803..2531895a 100644
--- a/koptreflow.h
+++ b/koptreflow.h
@@ -24,7 +24,9 @@
 #include "k2pdfopt.h"
 #include "context.h"
 
+K2PDFOPT_EXPORT
 void k2pdfopt_reflow_bmp(KOPTContext *kctx);
+K2PDFOPT_EXPORT
 void pixmap_to_bmp(WILLUSBITMAP *bmp, unsigned char *pix_data, int ncomp);
 
 #endif
diff --git a/willuslib/willus.h b/willuslib/willus.h
index 50c7fb27..e8673b5d 100644
--- a/willuslib/willus.h
+++ b/willuslib/willus.h
@@ -460,9 +460,13 @@ void bmp8_to_grey(WILLUSBITMAP *bmap);
 void bmp8_to_bw(WILLUSBITMAP *bmap,int thresh);
 int  bmp8_greylevel_convert(int r,int g,int b);
 #define bmp8_graylevel_convert(r,g,b) bmp8_greylevel_convert(r,g,b)
+K2PDFOPT_EXPORT
 void bmp_init(WILLUSBITMAP *bmap);
+K2PDFOPT_EXPORT
 int  bmp_alloc(WILLUSBITMAP *bmap);
+K2PDFOPT_EXPORT
 int  bmp_bytewidth(WILLUSBITMAP *bmp);
+K2PDFOPT_EXPORT
 unsigned char *bmp_rowptr_from_top(WILLUSBITMAP *bmp,int row);
 void bmp_crop(WILLUSBITMAP *bmp,int x0,int y0_from_top,int width,int height);
 void bmp_crop_ex(WILLUSBITMAP *dst,WILLUSBITMAP *src,int x0,int y0_from_top,int width,int height);
@@ -470,11 +474,13 @@ void bmp_rotate_fast(WILLUSBITMAP *dst,double degrees,int expand);
 int  bmp_rotate_right_angle(WILLUSBITMAP *bmp,int degrees);
 int  bmp_rotate_90(WILLUSBITMAP *bmp);
 int  bmp_rotate_270(WILLUSBITMAP *bmp);
+K2PDFOPT_EXPORT
 int  bmp_copy(WILLUSBITMAP *dest,WILLUSBITMAP *src);
 double bmp_jpeg_bytes_per_pixel(int quality);
 void bmp_flip_horizontal(WILLUSBITMAP *bmp);
 void bmp_flip_vertical(WILLUSBITMAP *bmp);
 int  bmp_bytewidth_win32(WILLUSBITMAP *bmp);
+K2PDFOPT_EXPORT
 void bmp_free(WILLUSBITMAP *bmap);
 int  bmp_read(WILLUSBITMAP *bmap,char *filename,FILE *out);
 void bmp24_reduce_size(WILLUSBITMAP *bmp,int mx,int my);
