diff --git a/willuslib/willus.h b/willuslib/willus.h
index d57d011f..4023e05b 100644
--- a/willuslib/willus.h
+++ b/willuslib/willus.h
@@ -236,6 +236,7 @@ typedef double  real;
 **
 */
 
+#if 0
 /*
 ** CMAKE handles the defines, not this source
 ** (Mod from Dirk Thierbach, 31-Dec-2013)
@@ -296,6 +297,9 @@ typedef double  real;
 #endif
 
 #endif /* USE_CMAKE */
+#else
+#include "willus_config.h"
+#endif
 /*
 ** Consistency check
 */
